### PR TITLE
feat(proof-gen): restart proof gen server when CC3 attestations stall (CSUB-2039)

### PR DESCRIPTION
**Pull request by DylanVerstraete**
## Summary:
1. Creates a new health condition. If we haven't seen a new attestation event in the last 5 minutes then the proof gen server is considered unhealthy. 
      a. While this condition might have false positives in the case of a stalled attestation chain, that doesn't really hurt anything.
      b. This acts as a strong failsafe preventing prolonged periods of down time.
2. If any of our health check conditions fail, we now respond to the health check with a 503 error. This will cause k8s to restart the pod. Previously we were returning 200's even when rpcs were broken.

Fix for [CSUB-2039](https://gluwa.atlassian.net/browse/CSUB-2039): when the CC3 `BlockAttested` event listener silently stops delivering events on cc3 devnet sepolia, every proof request starts failing with *"not yet attested"* while the proof gen server still reports itself as healthy. We add an attestation-liveness watchdog so a k8s liveness probe can restart the server and force a fresh CC3 subscription.

[CSUB-2039]: https://gluwa.atlassian.net/browse/CSUB-2039?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: proof-gen-api-server/src/services/continuity_service/mod.rs [hunk 1]
```diff
@@ -400,7 +417,19 @@ impl ContinuityService {
...
+    /// event is for a chain we don't serve, because a fresh event from CC3
+    /// still proves the listener is alive. The fix from CSUB-2039 relies on
+    /// this reset happening here so a stalled listener can be detected by the
+    /// liveness probe and the orchestrator can restart the server.
     pub async fn insert_attestation(&self, chain_key: u64, block_number: u64, digest: H256) {
+        // Reset liveness timer first — we want the timer to track "last
+        // attestation event observed from CC3", regardless of whether the
+        // event is for a chain this server happens to serve.
+        *self.last_attestation_event_at.write().await = Instant::now();
+
```
[the remark this answers is not in the document]
Comment (BradleyOlson64) [reply proof-gen-api-server/src/services/continuity_service/mod.rs:431]: Not actually strictly necessary to reset this on unserved chain events. It will work anyways

**Review by BradleyOlson64 — DISMISSED** [review]
LGTM. Imo this one is an important failsafe for preventing long outages due to proof gen server liveness issues.

File: proof-gen-api-server/src/networking/routes/health.rs [hunk 2]
```diff
@@ -48,17 +60,30 @@ pub async fn health_check(
...
-        cc3_rpc_connected: cc3_connected,
-        eth_rpc_connected: eth_connected,
-        total_proof_requests,
-        uptime_seconds: service.uptime_seconds(),
-    })
+    let http_status = if healthy {
+        StatusCode::OK
+    } else {
+        StatusCode::SERVICE_UNAVAILABLE
+    };
```
[the remark this answers is not in the document]
Comment (BradleyOlson64) [reply proof-gen-api-server/src/networking/routes/health.rs:76]: Was bad description

**Review by DylanVerstraete — COMMENTED** [review]
Why would we do this if https://github.com/gluwa/creditcoin3/pull/907 was merged

**Comment by BradleyOlson64** [comment]
@DylanVerstraete 
> Why would we do this if #907 was merged

Imo, we shouldn't assume that we caught every way the rpc connections and event streams can become unhealthy.

Dave seemed very in favor of having simple failsafes that cause prover restarts.

**Review by BradleyOlson64 — DISMISSED** [review]

**Comment by BradleyOlson64** [comment]
No K8s liveness probe currently set for proof gen server on cc3-testnet.

Could add one to the deployment with config like:

livenessProbe:
  httpGet:
    path: /api/v1/health
    port: 3100
  initialDelaySeconds: 600
  periodSeconds: 30
  timeoutSeconds: 5
  failureThreshold: 3
  
  This would give us a restart within 90 seconds of the first failed health check.
  
  Protoclaw suggested having a liveness `grace window` after restart programmed in code. 
  
  But I think it makes more sense to make use of the `initialDelaySeconds` field in the livenessProbe config.

File: proof-gen-api-server/src/services/continuity_service/helpers.rs [hunk 3]
```diff
@@ -729,6 +759,57 @@ mod tests {
...
+    #[tokio::test]
+    async fn check_attestation_event_timer_trips_after_timeout() {
+        // 0-duration timeout means "any elapsed time trips the check". This
+        // also exercises the boundary condition (>= timeout).
+        let svc = make_service_with_configuration(
+            Arc::new(FoundEthProvider),
+            DEFAULT_MAX_BATCH_SPAN,
+            std::time::Duration::from_secs(0),
+        )
+        .await;
```
[the remark this answers is not in the document]
Comment (BradleyOlson64) [reply proof-gen-api-server/src/services/continuity_service/helpers.rs:958]: Fixed

**Review by beqaabu — APPROVED** [review]

**Comment by DylanVerstraete** [comment]
Newer code deployed now should retry gracefully when RPC runs into any issue so closing this for now
